### PR TITLE
Approved roles

### LINKED ISSUE #18: Exclude `contributor` approvals
Related to https://github.com/ubiquibot/command-start-stop/issues/36

Any `contributor` role based approvals are invalid and should be excluded from any consideration in auto-merging. The only valid roles would be:

- `collaborator`
- `member`

_original [context](https://github.com/ubiquity/pay.ubq.fi/pull/290#issuecomment-2333291028)_

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #18

- filters reviews based on the `allowedRoles` previously used for assigning the timeout requirements
- added test and confirmed with local script on the _original context_ issue
- `yarn format` made a few changes
- added `package_manager` field to `package.json`

### CODE REVIEW

File: src/helpers/github.ts [hunk 1]
```diff
@@ -19,6 +19,8 @@ export interface Requirements {
   requiredApprovalCount: number;
 }
 
+const allowedRoles = ["COLLABORATOR", "MEMBER", "OWNER"];
```
Comment (gentlementlegen) [inline src/helpers/github.ts:22]: Maybe we could have these being configured through the configuration file itself? Some third parties might want to change the allowed reviewers.

File: src/types/plugin-inputs.ts [hunk 2]
```diff
@@ -55,13 +55,18 @@ export const reposSchema = T.Object(
...
+const rolesWithAuthority = T.Array(T.String(), { default: ["COLLABORATOR", "MEMBER", "OWNER"] })
+
 export const pluginSettingsSchema = T.Object({
   approvalsRequired: approvalsRequiredSchema,
   mergeTimeout: mergeTimeoutSchema,
   /**
    * The list of organizations or repositories to watch for updates.
    */
   repos: reposSchema,
+  rolesWithAuthority: T.Transform(rolesWithAuthority)
```
Comment (gentlementlegen) [inline src/types/plugin-inputs.ts:67]: `allowedReviewerRoles` might be more explicit here.

**Comment by gentlementlegen** [comment]
I think there is a logic issue. Take the following scenario:
In the configuration, I set `allowedReviewerRoles` to `["OWNER"]`. When the merge timeout will try to get the deadline requirements, a `MEMBER` would be considered as a `COLLABORATOR` instead of `CONTRIBUTOR` because `COLLABORATOR` is not in the `allowedReviewerRoles`.

Example run: https://github.com/Meniole/automated-merging/actions/runs/10748866022/job/29813121322#step:6:68

**Comment by Keyrxng** [comment]
> I think there is a logic issue. Take the following scenario: In the configuration, I set `allowedReviewerRoles` to `["OWNER"]`. When the merge timeout will try to get the deadline requirements, a `MEMBER` would be considered as a `COLLABORATOR` instead of `CONTRIBUTOR` because `COLLABORATOR` is not in the `allowedReviewerRoles`.
> 
> Example run: https://github.com/Meniole/automated-merging/actions/runs/10748866022/job/29813121322#step:6:68

I'm a little confused by what you are implying. I think you are suggesting that the deadline requirement array should be the same as it was before that way they'd be assigned `COLLABORATOR` and not the default `CONTRIBUTOR` in the event that the `allowedReviewerRoles` does not contain `COLLABORATOR`, is that right @gentlementlegen?

**Comment by gentlementlegen** [comment]
With the following configuration
```yml
  - uses:
    - plugin: ubiquity/automated-merging@development
      with:
        approvalsRequired:
          collaborator: 1
        mergeTimeout:
          collaborator: "2 minutes"
        repos:
          monitor: [ ]
          ignore: ["conversation-rewards"]
        allowedReviewerRoles: ["OWNER"]
```

Someone with the `COLLABORATOR` role would not be able to be considered to be counted within the reviewers, which is ok and intended. However if the collaborator is the one opening the pull-request, its timing for closing would be the ones for `contributor` instead of `collaborator` because the array considered to check the merge timeout is only `OWNER`. Hope it is clear enough.

**Comment by 0x4007** [comment]
What is an "owner" role? There should only be, basically: admin, assignee, collaborator, contributor.

**Comment by gentlementlegen** [comment]
`OWNER` is the creator of the repository. I took this as an arbitrary example, same would apply if I used `ADMIN`, that would be considered as a `CONTRIBUTOR` for merging timeout in that example.

File: package.json [hunk 3]
```diff
@@ -84,5 +84,6 @@
     "extends": [
       "@commitlint/config-conventional"
     ]
-  }
-}
+  },
+  "packageManager": "yarn@1.22.22"
```
Comment (0x4007) [inline package.json:88]: Can you fix your yarn

**Comment by Keyrxng** [comment]
> With the following configuration
> 
> ```yaml
>   - uses:
>     - plugin: ubiquity/automated-merging@development
>       with:
>         approvalsRequired:
>           collaborator: 1
>         mergeTimeout:
>           collaborator: "2 minutes"
>         repos:
>           monitor: [ ]
>           ignore: ["conversation-rewards"]
>         allowedReviewerRoles: ["OWNER"]
> ```
> 
> Someone with the `COLLABORATOR` role would not be able to be considered to be counted within the reviewers, which is ok and intended. However if the collaborator is the one opening the pull-request, its timing for closing would be the ones for `contributor` instead of `collaborator` because the array considered to check the merge timeout is only `OWNER`. Hope it is clear enough.

Not sure how else to address the scenario that you raised apart from retaining your original implementation and hardcoding the timeout roles.

1. Potential user errors using the same `allowedReviewerRoles` item and additional unneeded complexity adding a new config item for this array separately.
2. I figured that specific aspect is author-defined like the way `skipBotsEvents` should be as oppose to author defined.

> Can you fix your yarn
I can't find your comment rn but didn't we agree that we'd standardize this? No harm done including it but potential room for error without it. I stated it as an addition in the spec body too.

**Comment by 0x4007** [comment]
Don't understand your message. What's the problem?

**Comment by gentlementlegen** [comment]
Seems that it works properly, my last QA: https://github.com/Meniole/automated-merging/actions/runs/10828231151/job/30043117944#step:6:68

**Comment by Keyrxng** [comment]
> Don't understand your message. What's the problem?

I don't understand you message now lmao.

- If you mean the issue raised by mentlegen, I was unsure myself as I expected `allowedReviewerRoles` to always be the same for fetching timeouts and fetching authorized PR approvals, they appear like they'd always rely on the same roles. I scratched my head with a solve and revert to original was the only guaranteed way to avoid the scenario raised.
- If you mean `packageManager` then the [docs](https://nodejs.org/api/corepack.html) detail things better than I do

> This feature simplifies two core workflows:

> It eases new contributor onboarding, since they won't have to follow system-specific installation processes anymore just to have the package manager you want them to.

> It allows you to ensure that everyone in your team will use exactly the package manager version you intend them to, without them having to manually synchronize it each time you need to make an update.

**Review by gentlementlegen — APPROVED** [review]

**Comment by 0x4007** [comment]
@gentlementlegen why did UbiquityOS handle the merge?

**Comment by gentlementlegen** [comment]
@0x4007 I noticed that the configuration was set to 5 minutes for collaborators, probably for testing, I corrected that afterwards so it doesn't happen again. Maybe we could do sentiment analysis and eventually merge lol

**Comment by 0x4007** [comment]
```diff
+ ~ * checking vibes, please wait * ~
```

**Pull request merged**